server/vulnerabilities/nvd/sync/cve_syncer.go

```diff
@@ -0,0 +1,727 @@
+			legacyCVE := convertAPI20CVEToLegacy(cve, s.logger)
+			newLegacyCVEs[legacyCVE.CVE.CVEDataMeta.ID] = legacyCVE
+		}
+
+		// Update existing CVEs with the lastest updates.
+		for i, storedCVE := range storedCVEFeed.CVEItems {
+			if newLegacyCVE, ok := newLegacyCVEs[storedCVE.CVE.CVEDataMeta.ID]; ok {
+				storedCVEFeed.CVEItems[i] = newLegacyCVE
+				delete(newLegacyCVEs, storedCVE.CVE.CVEDataMeta.ID)
+			}
+		}
+
+		// Add any new CVEs.
+		for _, cve := range newLegacyCVEs {
+			storedCVEFeed.CVEItems = append(storedCVEFeed.CVEItems, cve)
+		}
+
+		// Store the file for the year.
+		if err := storeCVEsInLegacyFormat(s.dbDir, cvesInYear.year, storedCVEFeed); err != nil {
+			return err
+		}
+	}
+
+	// Update the lastModStartDate for the next synchronization.
+	if err := s.writeLastModStartDateFile(lastModStartDate); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// writeLastModStartDateFile writes the lastModStartDate to a file in the local DB directory.
+func (s *CVE) writeLastModStartDateFile(lastModStartDate string) error {
+	if err := os.WriteFile(
+		s.lastModStartDateFilePath(),
+		[]byte(lastModStartDate),
+		constant.DefaultWorldReadableFileMode,
+	); err != nil {
+		return err
+	}
+	return nil
+}
+
+// cvesInYear contains a list of CVEs for a specific year.
+type cvesInYear struct {
+	year int
+	cves []nvdapi.CVEItem
+}
+
+// httpClient wraps an http.Client to allow for debug and setting a request context.
+type httpClient struct {
+	*http.Client
+	ctx context.Context
+
+	debug bool
+}
+
+// Do implements common.HTTPClient.
+func (c *httpClient) Do(request *http.Request) (*http.Response, error) {
+	if c.debug {
+		fmt.Fprintf(os.Stderr, "%+v\n", request)
+	}
+
+	response, err := c.Client.Do(request.WithContext(c.ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	if c.debug {
+		bodyBytes, _ := io.ReadAll(response.Body)
+		response.Body.Close()
+		response.Body = io.NopCloser(bytes.NewBuffer(bodyBytes))
+		fmt.Fprintf(os.Stderr, "%+v %s\n%s\n", response, err, bodyBytes)
+	}
+
+	return response, err
+}
+
+// getHTTPClient returns common.HTTPClient to be used by nvdapi methods.
+func (s *CVE) getHTTPClient(ctx context.Context, debug bool) common.HTTPClient {
+	return &httpClient{
+		Client: s.client,
+		ctx:    ctx,
+
+		debug: debug,
+	}
+}
+
+// sync performs requests to the NVD https://services.nvd.nist.gov/rest/json/cves/2.0 service to get CVE information.
+// It returns the fetched CVEs and the lastModStartDate to use on a subsequent sync call.
+//
+// If lastModStartDate is nil, it performs the initial (full) synchronization of ALL CVEs.
+// If lastModStartDate is set, then it fetches updates since the last sync call.
+//
+// Reference: https://nvd.nist.gov/developers/api-workflows.
+func (s *CVE) sync(ctx context.Context, lastModStartDate *string) (cves []cvesInYear, newLastModStartDate string, err error) {
+	var (
+		totalResults   = 1
+		cvesByYear     = make(map[int][]nvdapi.CVEItem)
```

Comment:
Yes, there could be memory issues here.

The reason I loaded them all is because I was not making any assumptions around the order in which the CVEs are returned. But maybe NVD returns them in order so the moment we get a CVE for year X we can store year X-1.

I will try the optimization and keep this thread posted.